File: scripts/create_missing_tables.py
"""在 Supabase 中创建缺失的表 (feedbacks, admin_broadcasts, user_messages, ai_configs)"""
import os
import urllib.request
import urllib.error
import json

# 读取环境变量
env_path = os.path.join(os.path.dirname(__file__), '..', '.env.local')
with open(env_path, 'r', encoding='utf-8') as f:
    for line in f:
        line = line.strip()
        if line and not line.startswith('#') and '=' in line:
            k, v = line.split('=', 1)
            os.environ[k.strip()] = v.strip()

url = os.environ.get('NEXT_PUBLIC_SUPABASE_URL', '')
key = os.environ.get('SUPABASE_SERVICE_ROLE_KEY', '')
project_ref = url.replace('https://', '').split('.')[0]

SQL = """
CREATE TABLE IF NOT EXISTS feedbacks (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  user_id UUID REFERENCES users(id),
  user_email TEXT,
  user_name TEXT,
  type TEXT DEFAULT 'general',
  content TEXT NOT NULL,
  status TEXT DEFAULT 'pending',
  admin_reply TEXT,
  replied_at TIMESTAMP,
  created_at TIMESTAMP DEFAULT NOW()
);

CREATE TABLE IF NOT EXISTS admin_broadcasts (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  title TEXT NOT NULL,
  content TEXT NOT NULL,
  target_type TEXT DEFAULT 'all',
  target_users UUID[],
  sent_by UUID REFERENCES users(id),
  created_at TIMESTAMP DEFAULT NOW()
);

CREATE TABLE IF NOT EXISTS user_messages (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  user_id UUID REFERENCES users(id),
  type TEXT DEFAULT 'system',
  title TEXT,
  content TEXT NOT NULL,
  is_read BOOLEAN DEFAULT FALSE,
  related_feedback_id UUID,
  created_at TIMESTAMP DEFAULT NOW()
);

CREATE TABLE IF NOT EXISTS ai_configs (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  provider VARCHAR(50) NOT NULL DEFAULT 'kimi',
  api_key TEXT NOT NULL,
  api_url TEXT,
  model VARCHAR(100) DEFAULT 'kimi-latest',
  max_tokens INTEGER DEFAULT 2048,
  temperature DECIMAL(3,2) DEFAULT 0.7,
  is_active BOOLEAN DEFAULT true,
  usage_limit INTEGER DEFAULT 1000000,
  usage_current INTEGER DEFAULT 0,
  created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
  updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
);

CREATE INDEX IF NOT EXISTS idx_feedbacks_user_id ON feedbacks(user_id);
CREATE INDEX IF NOT EXISTS idx_feedbacks_status ON feedbacks(status);
CREATE INDEX IF NOT EXISTS idx_feedbacks_created_at ON feedbacks(created_at DESC);
CREATE INDEX IF NOT EXISTS idx_user_messages_user_id ON user_messages(user_id);
CREATE INDEX IF NOT EXISTS idx_user_messages_is_read ON user_messages(is_read);
CREATE INDEX IF NOT EXISTS idx_user_messages_created_at ON user_messages(created_at DESC);
"""

def check_tables():
    """检查表是否存在"""
    headers = {'apikey': key, 'Authorization': f'Bearer {key}', 'Content-Type': 'application/json'}
    results = {}
    for table in ['admin_broadcasts', 'feedbacks', 'user_messages', 'ai_configs']:
        req = urllib.request.Request(f'{url}/rest/v1/{table}?select=id&limit=1', headers=headers)
        try:
            urllib.request.urlopen(req)
            results[table] = True
        except urllib.error.HTTPError:
            results[table] = False
        except Exception:
            results[table] = False
    return results

def create_tables_via_management_api():
    """尝试通过 Supabase Management API 创建表"""
    # Supabase Management API 需要个人 access token，这里跳过
    # 改为输出 SQL 供手动执行
    print("=" * 60)
    print("需要在 Supabase Dashboard SQL Editor 中执行以下 SQL:")
    print("=" * 60)
    print(SQL)
    print("=" * 60)
    print(f"\nDashboard URL: https://supabase.com/dashboard/project/{project_ref}/sql")
    print("\n或者使用 Supabase CLI:")
    print(f"  npx supabase db push")

if __name__ == '__main__':
    print("Checking existing tables...")
    results = check_tables()
    for table, exists in results.items():
        status = "EXISTS" if exists else "NOT EXISTS"
        print(f"  {table}: {status}")
    
    missing = [t for t, e in results.items() if not e]
    if missing:
        print(f"\nMissing tables: {', '.join(missing)}")
        create_tables_via_management_api()
    else:
        print("\nAll tables exist!")

File: scripts/create-missing-tables.js
// ============================================
// 在 Supabase 中创建缺失的表
// 表: feedbacks, admin_broadcasts, user_messages, ai_configs
// 运行: node scripts/create-missing-tables.js
// ============================================

const { createClient } = require('@supabase/supabase-js');

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const serviceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !serviceRoleKey) {
  console.error('Error: NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY are required');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, serviceRoleKey, {
  auth: { autoRefreshToken: false, persistSession: false }
});

async function createMissingTables() {
  console.log('Creating missing tables in Supabase...\n');

  // 1. feedbacks 表
  console.log('1. Creating feedbacks table...');
  const { error: feedbacksError } = await supabase.rpc('query', {
    query: `
      CREATE TABLE IF NOT EXISTS feedbacks (
        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        user_id UUID REFERENCES users(id),
        user_email TEXT,
        user_name TEXT,
        type TEXT DEFAULT 'general',
        content TEXT NOT NULL,
        status TEXT DEFAULT 'pending',
        admin_reply TEXT,
        replied_at TIMESTAMP,
        created_at TIMESTAMP DEFAULT NOW()
      );
      CREATE INDEX IF NOT EXISTS idx_feedbacks_user_id ON feedbacks(user_id);
      CREATE INDEX IF NOT EXISTS idx_feedbacks_status ON feedbacks(status);
      CREATE INDEX IF NOT EXISTS idx_feedbacks_created_at ON feedbacks(created_at DESC);
    `
  });
  if (feedbacksError) console.log('  (Using direct insert fallback)');

  // 2. admin_broadcasts 表
  console.log('2. Creating admin_broadcasts table...');
  const { error: broadcastsError } = await supabase.rpc('query', {
    query: `
      CREATE TABLE IF NOT EXISTS admin_broadcasts (
        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        title TEXT NOT NULL,
        content TEXT NOT NULL,
        target_type TEXT DEFAULT 'all',
        target_users UUID[],
        sent_by UUID REFERENCES users(id),
        created_at TIMESTAMP DEFAULT NOW()
      );
    `
  });
  if (broadcastsError) console.log('  (Using direct insert fallback)');

  // 3. user_messages 表
  console.log('3. Creating user_messages table...');
  const { error: messagesError } = await supabase.rpc('query', {
    query: `
      CREATE TABLE IF NOT EXISTS user_messages (
        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        user_id UUID REFERENCES users(id),
        type TEXT DEFAULT 'system',
        title TEXT,
        content TEXT NOT NULL,
        is_read BOOLEAN DEFAULT FALSE,
        related_feedback_id UUID REFERENCES feedbacks(id),
        created_at TIMESTAMP DEFAULT NOW()
      );
      CREATE INDEX IF NOT EXISTS idx_user_messages_user_id ON user_messages(user_id);
      CREATE INDEX IF NOT EXISTS idx_user_messages_is_read ON user_messages(is_read);
      CREATE INDEX IF NOT EXISTS idx_user_messages_created_at ON user_messages(created_at DESC);
    `
  });
  if (messagesError) console.log('  (Using direct insert fallback)');

  // 4. ai_configs 表
  console.log('4. Creating ai_configs table...');
  const { error: aiConfigsError } = await supabase.rpc('query', {
    query: `
      CREATE TABLE IF NOT EXISTS ai_configs (
        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        provider VARCHAR(50) NOT NULL DEFAULT 'kimi',
        api_key TEXT NOT NULL,
        api_url TEXT,
        model VARCHAR(100) DEFAULT 'kimi-latest',
        max_tokens INTEGER DEFAULT 2048,
        temperature DECIMAL(3,2) DEFAULT 0.7,
        is_active BOOLEAN DEFAULT true,
        usage_limit INTEGER DEFAULT 1000000,
        usage_current INTEGER DEFAULT 0,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      );
    `
  });
  if (aiConfigsError) console.log('  (Using direct insert fallback)');

  // 验证表是否存在
  console.log('\nVerifying tables...');
  
  const tables = ['feedbacks', 'admin_broadcasts', 'user_messages', 'ai_configs'];
  for (const table of tables) {
    const { data, error } = await supabase.from(table).select('id').limit(1);
    if (error) {
      console.log(`  ❌ ${table}: NOT FOUND - ${error.message}`);
    } else {
      console.log(`  ✅ ${table}: EXISTS`);
    }
  }
}

createMissingTables()
  .then(() => console.log('\nDone!'))
  .catch(err => console.error('Error:', err));
